Preferred Email/Phone Translation Issue Fix

skosgi1 opened the pull request:
# Critical Changes

# Changes
## Updates to the Preferred Phone Functionality
The enhanced Preferred Phone functionality released in update 1.84 can now be enabled in orgs that translate Catalan, Spanish, or French.
- If you are an existing EDA customer who wants to enable the feature in an org that translates any of these languages, go to EDA Settings | Accounts and Contacts and check the "Enable Enhanced Preferred Phone Functionality" checkbox.
- If you're installing in a new org that translates any of these languages, the enhanced Preferred Phone logic is enabled by default, so you don't need to change the new setting. However, we do recommend that you uncheck "Disable Preferred Phone enforcement" if you want to ensure that Contacts have a Preferred Phone value when more than one Phone field is specified on the Contact record.
- If you're translating in any other languages and want to use this feature, use the translation workbench to update Preferred Phone and Preferred Email picklist values to match the translated field labels.

**Note:** Existing customers might notice that some of EDA's unit tests fail after this upgrade. It's not required that EDA unit tests pass in customer orgs, but if you want them to pass, updating the Contact's Preferred Phone picklist values in your org to "Home Phone," "Mobile Phone," "Other Phone," and "Work Phone" should fix failures related to the upgrade.

# Issues Closed
Fixes #946 

# New Metadata

# Deleted Metadata

# Testing Notes

Testing Preferred Phone:

1. Make sure Enhanced Preferred Phone Functionality is checked : EDA Settings → Accounts & Contacts → Enable Enhanced Preferred Phone Functionality:
2. Make sure "CON_PreferredPhone_TDTM" trigger handler is active in "Trigger handlers" Tab.
3. Create a Contact with Preferred Phone as ‘Home’ and enter a value in the field ‘Home Phone’ and try saving the record. The record saves.
4. Change the language from user → User record →  Change Language to Spanish OR  
    click your avatar on top right → Settings → Language and Time zone → Español
5. Now create a contact with ‘Teléfono elegido’ as ‘Teléfono particular’  and enter a value for the field ‘Teléfono particular’  and try to save the record, the record should be saved.

Testing Preferred Email:

1.  Make sure "CON_PreferredEmail_TDTM" trigger handler is active in "Trigger handlers" Tab.
2.Create a contact with Preferred Email as "Work" and enter a value in the field "Work Email: and try saving the record. The record saves.
3.  Change the language from user → User record →  Change Language to Spanish OR  
    click your avatar on top right → Settings → Language and Time zone → Español
4. Create a Contact with "Correo electrónico elegido" as "Correo electrónco laboral" and enter a value in "Correo electrónico laboral" and try to save the record, the record should be saved.

Test Push Upgrade for Old Customers vs New Install (New Customers)
Test Two+ Phones ("Tell us which is preferred.")
Test Two Emails ("Tell us which is preferred.")
Test Run Cleanup

Note : For old customers, the picklist values must match the field label values (except for in English.) For new customers, the picklist values must match the field label values for all 4 languages (English, French, Spanish, Catalan.)

ceiroa commented on src/classes/CON_Phone.cls:
```diff
@@ -314,11 +314,28 @@ public with sharing class CON_Phone {
+             
+            if (!preferredPhoneValues.contains(lbl)) {
+                this.prefLabel = (preferredPhoneMappings.containsKey(lbl)) ? preferredPhoneMappings.get(lbl) : lbl;
+                System.debug('11111111 lbl'+this.prefLabel);
```
Remove debug statements?

skosgi1 replied:
Oops missed that. Removed it now!

ceiroa commented on src/classes/CON_Email.cls:
```diff
@@ -281,11 +281,24 @@ public class CON_Email {
-
+                     
+            String namespace = UTIL_Namespace.getNamespace();
+            String prefix = (String.isNotBlank(namespace)) ? namespace + '__' : '';
```
We also support custom fields (user would add a value to the picklist and the field to the layout). Would that functionality be affected?

skosgi1 replied:
That functionality would work if the label and picklist have an exact match.

ceiroa replied:
So the prefix won't be added for fields that are not in the package? 

Ideally we would test this still works, but I guess the only way to test it is after merging (with a beta). Unfortunately namespaced scratch orgs don't help here...

skosgi1 replied:
Hmm we are appending the prefix for only "PreferredPhone__c" and "Preferred_Email__c" fields which are packaged but we are not appending it to the actual phone/email fields. Please let me know if I that was your question.

ceiroa replied:
Ah, sorry. My bad. I looked at this too quickly. Yes, we are applying the prefix only to the field, not to the picklist values (which I guess don't need it anyway). Thanks for bearing with me, Samsheeth!

ceiroa commented on src/classes/CON_Phone.cls:
```diff
@@ -314,11 +314,28 @@ public with sharing class CON_Phone {
-            this.prefLabel = (preferredPhoneMappings.containsKey(lbl)) ? preferredPhoneMappings.get(lbl) : lbl;
+            
+            String namespace = UTIL_Namespace.getNamespace();
+            String prefix = (String.isNotBlank(namespace)) ? namespace + '__' : '';
```
Same question as above regarding prefix.

skosgi1 replied:
That functionality would work if the label and picklist have an exact match.

ceiroa commented on src/objectTranslations/Contact-ca.objectTranslation:
```diff
@@ -3473,36 +3473,64 @@
         <picklistValues>
             <masterLabel>Home</masterLabel>
-            <translation>Llar</translation>
+            <translation>Home Phone</translation>
```
We should use the Catalán translation for all these, not English (available in the fields with the same name).

skosgi1 replied:
I think since this is a Platform Only language some fields are not being translated, so I updated the picklist translations corresponding to those fields to English.

robertbwatson replied:
They were previously translated though, right? So that might be a bit confusing. Is it because the Home, Mobile, and Other are not translated in the UI, and so if we translated the picklist, the functionality would still not work for them?

skosgi1 replied:
Correct!

robertbwatson replied:
Interesting. Thoughts on this one, @ceiroa? We're essentially taking away translations... but... this wouldn't be working for old or new customers anyway, since the actual fields themselves are not translated by Salesforce.

So, Catalan customers should translate both the Preferred Phone picklist and the standard field values if they need this to work in Catalan.

robertbwatson commented on src/classes/CON_Email.cls:
```diff
@@ -281,11 +281,24 @@ public class CON_Email {
             this.label = lbl;
             this.apiName = api;
-            this.prefLabel = (preferredEmailMappings.containsKey(lbl)) ? preferredEmailMappings.get(lbl) : lbl;
+            if (!preferredEmailValues.contains(lbl)) {
```
@skosgi1 Some comments explaining this might be ideal. Something like: "Use the old preferred email mappings if the picklist values do not contain the field label"?

Or, maybe update the comment on line 278 "Store default email field mappings to preferred email picklist values prior to EDA 1.86."

skosgi1 replied:
Good point. Updated it.

robertbwatson commented on src/classes/CON_Phone.cls:
```diff
@@ -314,11 +314,25 @@ public with sharing class CON_Phone {
+                preferredPhoneValues.add(eachValue.getValue());
+            }
+             
+            if (!preferredPhoneValues.contains(lbl)) {
```
Same comment as above re: adding an explanation here or updating the comment on line 310.

skosgi1 replied:
Updated it.

robertbwatson commented on src/objectTranslations/Contact-ca.objectTranslation:
```diff
@@ -3473,36 +3473,64 @@
         <picklistValues>
             <masterLabel>Alternate</masterLabel>
-            <translation>Alternatiu</translation>
+            <translation>orreu electrònic alternatiu</translation>
```
Missing a C at the beginning, I think...

skosgi1 replied:
Good catch! Updated it.

robertbwatson commented on src/objectTranslations/Contact-es.objectTranslation:
```diff
@@ -3473,19 +3473,35 @@
         <picklistValues>
             <masterLabel>Other</masterLabel>
-            <translation>Otros</translation>
+            <translation>Otro teléfono</translation>
```
I'm worried in all of these translations that maybe we're guessing and we should actually leave this up to the translations team or at least run it by them first. 

For example, why is Other / Other Phone "Otro telefono", whereas all of the others have "Telefono ..." first?

(I don't know any other languages so it's possible that this is correct, but I'm questioning it...)

Thoughts, @skosgi1 @ceiroa?

skosgi1 replied:
Yeah looks like one is "other" and the other one translates to "another" ? but they have to match the field labels anyway for the functionality to work, so not sure about the inconsistency here...

ceiroa replied:
Well, we don't have to guess if we are taking the value from the label. 

Also, yeah, it is correct that in this case "teléfono" goes after and the others have it first.

robertbwatson replied:
@skosgi1 haha! I'm sorry. I forgot that we were mapping these to the actual field labels, which *are* translated already by Salesforce. Nevermind!

robertbwatson commented on src/objectTranslations/Contact-fr.objectTranslation:
```diff
@@ -3494,15 +3510,27 @@
         <picklistValues>
             <masterLabel>Work</masterLabel>
-            <translation>Professionnel</translation>
+            <translation>E-mail alternatif</translation>
```
This seems wrong also.

skosgi1 replied:
Good catch! Updated it.

robertbwatson commented on src/objectTranslations/Contact-fr.objectTranslation:
```diff
@@ -3494,15 +3510,27 @@
         <picklistValues>
             <masterLabel>Alternate</masterLabel>
-            <translation>Autre</translation>
+            <translation>E-mail alternatif</translation>
```
Why was the translation changed completely here?

skosgi1 replied:
I think "Autre" translate to "Other" and the change translates to "Alternate Email". It matches the field label also.

robertbwatson replied:
Nevermind - I forgot that you were just mapping them to the actual field labels

robertbwatson commented on robot/EDA/resources/locators.py:
```diff
@@ -144,7 +144,7 @@
     "new_account": "//span[@title='New Account']",
     "new_account_next_button": "//button[contains(@class, 'slds-button')]//span[@class=' label bBody' and text()='Next']",
-    "new_account_name": "//input[@class='input uiInput uiInputText uiInput--default uiInput--input']",
+    "new_account_name": "//label/span[text()='Account Name']/following-sibling::span/following::input[1]",
```
@dashattuck Same comment that I made earlier in another open PR - if this is being handled in #968, then let's remove it from here. I'd prefer to keep Robot changes separate from other work items.

robertbwatson replied:
Reverted, resolving.

lifewithryan commented:
> @robertbwatson @skosgi1 Sorry, for some reason I cannot reply inline - I still don't get how exactly Platform-only languages work. The [documentation](https://help.salesforce.com/articleView?id=faq_getstart_what_languages_does.htm&type=5) says "_[...]when you select a platform-only language, all standard Salesforce labels default to English or, in select cases, to an end-user or fully supported language._" I guess that means we should leave it in English? I don't get how they are able to provide a translation in Catalán, but it does sound like having it in English is the right thing to do...

@ceiroa this is something we came across in testing when they translated EDA. Some of the bits just simply don't translate. However I remember it being very unclear as to why/what/when.

ceiroa commented:
Alrighty. English it is then.

skosgi1 commented:
> Alrighty. English it is then.

In this case only work phone is translated and the rest of the phone fields/ picklist values are in English. So 1 picklist is translated into Catalan and 3 are not.

ceiroa commented:
Scratching my head on this one again, @skosgi1. Why is "work phone" translated? How is it different from the others? Should we just not translate any of them? It seems strange that that one is translated and not the others...

skosgi1 commented:
> Scratching my head on this one again, @skosgi1. Why is "work phone" translated? How is it different from the others? Should we just not translate any of them? It seems strange that that one is translated and not the others...

@ceiroa In this case only "Work Phone" is custom fields so we have it in the translation file, other phone fields are standard fields and this language being platform only - as per what I read not all standard fields get translated.

ceiroa commented:
@skosgi1 Ah, got you. I didn't realize the other phone fields were standard. Ok, then we should be good to keep work phone in English. Thanks!

robertbwatson commented:
> > Scratching my head on this one again, @skosgi1. Why is "work phone" translated? How is it different from the others? Should we just not translate any of them? It seems strange that that one is translated and not the others...
> 
> @ceiroa In this case only "Work Phone" is custom field so we have it in the translation file, other phone fields are standard fields and this language being platform only - as per what I read not all standard fields get translated.

Right - and for email functionality, Alternate Email, University Email, and Work Email are all custom fields that we've added to EDA, so we've translated them ourselves. Which means that picklist is correctly translated to Catalan. But until Salesforce supports Catalan as a core language, Home, Other, and Mobile phone fields remain in English.

skosgi1 commented:
> @skosgi1 Ah, got you. I didn't realize the other phone fields were standard. Ok, then we should be good to keep work phone in English. Thanks!

Hmm all the other picklists would be in English, but "Work Phone" would be in Catalan. Since only the field "Work Phone" is translated into Catalan. "Home Phone" , "Mobile Phone", "Other Phone" picklist values will be in English and "Work Phone" value will be in Catalan. If we keep "Work Phone" in English the functionality would not work for Catalan since the field Work Phone is now in Catalan.

robertbwatson commented:
@ceiroa FYI I've assigned this to you - since you requested changes it needs to be approved by you before it's moved to Dina.

skosgi1 commented:
@robertbwatson @ceiroa I have made the changes to the code to use API values instead of labels, can you give it a quick review when you get a moment? Thank you!

ceiroa commented on src/classes/CON_Phone.cls:
```diff
@@ -324,14 +324,17 @@ public with sharing class CON_Phone {
             for (Schema.PicklistEntry eachValue: preferredPhonePicklist) {
-                preferredPhoneValues.add(eachValue.getValue());
+                PreferredLabelToApi.put(eachValue.getLabel(),eachValue.getValue());
             }
```
@skosgi1 ~ Just to confirm, will `eachValue.getValue()` get the picklist's element API value?

lifewithryan replied:
That's the way I read this as well.

skosgi1 replied:
correct! getValue() is the API and getLabel() is the field label.

ceiroa commented on src/classes/CON_Phone.cls:
```diff
@@ -324,14 +324,17 @@ public with sharing class CON_Phone {
                 this.prefLabel = (preferredPhoneMappings.containsKey(lbl)) ? preferredPhoneMappings.get(lbl) : lbl;
             } else {            
-                 this.prefLabel = lbl;
+                 this.prefLabel = PreferredLabelToApi.get(lbl);
```
Also for my understanding; calling `PreferredLabelToApi.get(lbl)` will return the API value, not the label, right?

skosgi1 replied:
Correct!

ceiroa replied:
Do we care that the value is getting assigned to `prefLabel` even though it's the API name, or are we ok with that?

skosgi1 replied:
Hmm not sure if I understood your question correctly @ceiroa, We're assigning the value of the picklist API corresponding to picklist label that matches with the field label. Please let me know if it doesn't answer your question.

robertbwatson replied:
If I understand Carlos correctly, he's just concerned that the variable name "prefLabel" is a little misleading since it might actually be the API name.

I agree it is a little confusing. The comments are good here and in CON_Email.cls but I wonder if it would be nice to explain the "why" in the comments. The comments explain the code but don't necessarily explain why it is done this way.

skosgi1 commented:
@robertbwatson I updated the code as we discussed earlier. Can you take a quick pass it again? thx.

lifewithryan commented:
> @robertbwatson I updated the code as we discussed earlier. Can you take a quick pass it again? thx.

@robertbwatson @skosgi1 -- I've pulled this down too as I want to run through some things but was in a meeting so won't get a chance to finish until tomorrow. First skim looks good. Looks like there may be potential in the future to make the new method you created slightly more generic/flexible so that both email and phone could use it, but that's not anything I would hold up a merge for. more like a tech debt future task.

skosgi1 commented:
> > @robertbwatson I updated the code as we discussed earlier. Can you take a quick pass it again? thx.
> 
> @robertbwatson @skosgi1 -- I've pulled this down too as I want to run through some things but was in a meeting so won't get a chance to finish until tomorrow. First skim looks good. Looks like there may be potential in the future to make the new method you created slightly more generic/flexible so that both email and phone could use it, but that's not anything I would hold up a merge for. more like a tech debt future task.

@lifewithryan Thanks for the quick feedback. I added the new generic method to UTIL_Describe.cls now.

robertbwatson commented on src/classes/CON_Phone.cls:
```diff
@@ -183,6 +183,16 @@ public with sharing class CON_Phone {
+
+    /*******************************************************************************************************
+    * @description Returns a map with Preferred Phone picklist entry's label as key and API as value
+    * @retun Map<String, String> A key value pair of Preferred Phone picklist entrys's label and API respectively
```
Nit: return

robertbwatson commented on src/classes/CON_Email.cls:
```diff
@@ -162,6 +162,16 @@ public class CON_Email {
 
+    /*******************************************************************************************************
+    * @description Returns a map with Preferred Email picklist entry's label as key and API as value
+    * @retun Map<String, String> A key value pair of Preferred Email picklist entry's label and API respectively
```
Nit: return

robertbwatson commented on src/classes/CON_Phone.cls:
```diff
@@ -183,6 +183,16 @@ public with sharing class CON_Phone {
+    * @description Returns a map with Preferred Phone picklist entry's label as key and API as value
+    * @retun Map<String, String> A key value pair of Preferred Phone picklist entrys's label and API respectively
+    */
+    private static Map<String, String> createPreferredLabelToAPIMap() {             
```
I'm sorry for holding up the QA on this and adding a last minute suggestion (& also sorry for going AFK immediately after our meeting today!)

When I had mentioned caching performance enhancements here, I meant that it would be nice to only generate the map once. This will be especially useful if we have a large number of contacts that we're iterating over at once.

I won't hold up the review for this, but a future enhancement here could be something like:
```
private static Map<String, String> labelToAPIMap;
private static Map<String, String> getPreferredLabelToAPIMap() {
    if (labelToAPIMap == null) {
        // code goes here
        labelToAPIMap = UTIL_Describe...
    }
    return labelToAPIMap;
}
```

In this example, the map is only generated once and subsequent calls to `getPreferredLabelToAPIMap` return the stored value.

skosgi1 replied:
Thanks for the clarification @robertbwatson. I made the change, can you take another quick look? Thx.

robertbwatson replied:
Perfect @skosgi1 -- thanks!
